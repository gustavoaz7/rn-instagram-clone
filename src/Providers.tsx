import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider } from 'react-redux';
import { theme as stylesTheme } from './styles/theme';
import { store as reduxStore } from './redux/store';

type TProvidersProps = {
  children: ReactNode;
  store?: typeof reduxStore;
  theme?: typeof stylesTheme;
};

export function Providers({
  children,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <Provider store={store}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </Provider>
  );
}
