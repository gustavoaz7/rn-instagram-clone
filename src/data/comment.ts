import { TComment } from '../types';

export const generateComment = (): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
});
