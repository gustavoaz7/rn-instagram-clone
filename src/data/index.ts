import type {
  TLike,
  TComment,
  TPostMedia,
  TPost,
  TOwner,
  TStory,
  TStoryMedia,
  TProfile,
  TTappableObject,
} from '../types';

export const generateMockOwner = (): TOwner => ({
  id: `${Math.random()}`,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
});

export const generateMockLike = (): TLike => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
});

export const generateMockComment = (likesQty = 1): TComment => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
    username: 'Vivienne_Doyle',
  },
  createdAt: 1620488304326,
  text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  viewerHasLiked: false,
});

export const generateMockTappableObject = ({
  type = 'mention',
  text = 'random',
  x = 0.1,
  y = 0.2,
  scale = 1.1,
  rotation = 300,
}: Partial<TTappableObject> = {}): TTappableObject => ({
  id: `${Math.random()}`,
  type,
  text,
  x,
  y,
  scale,
  rotation,
});

export const generateMockPostMedia = ({ index = 0 } = {}): TPostMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
});

export const generateMockPost = ({
  mediasQty = 1,
  likesQty = 3,
  commentsQty = 1,
} = {}): TPost => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  createdAt: 1620488304326,
  medias: [...Array(mediasQty)].map(generateMockPostMedia),
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  previewLikes: {
    count: likesQty,
    likes: [...Array(likesQty)].map(generateMockLike),
  },
  previewComments: {
    count: commentsQty,
    comments: [...Array(commentsQty)].map(generateMockComment),
  },
  location: 'Daronville',
  viewerHasLiked: false,
});

export const generateMockStoryMedia = ({
  viewersQty = 0,
  index = 0,
} = {}): TStoryMedia => ({
  id: `${Math.random()}`,
  url: `http://placeimg.com/640/480/cats?${index}`,
  owner: generateMockOwner(),
  tappableObjects: [],
  previewViewers: {
    count: viewersQty,
    viewers: [...Array(viewersQty)].map(generateMockOwner),
  },
  takenAt: 1620488304326,
  expiresAt: 1620488304326 + 3600 * 24,
});

export const generateMockStory = ({
  mediasQty = 3,
  viewersQty = 0,
} = {}): TStory => ({
  id: `${Math.random()}`,
  owner: generateMockOwner(),
  expiresAt: 0,
  latestMediaAt: 0,
  seenAt: 0,
  medias: [...Array(mediasQty)].map((_, index) =>
    generateMockStoryMedia({ viewersQty, index }),
  ),
});

export const generateMockProfile = (): TProfile => ({
  bio: 'animi eum earum aut explicabo',
  followCount: 47,
  followedByCount: 50,
  followedByViewer: false,
  followsViewer: false,
  fullName: 'Gustavo Az',
  postsCount: 20,
  profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
  username: 'gustavoaz7',
  story: generateMockStory(),
});
