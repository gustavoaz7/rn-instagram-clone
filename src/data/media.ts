import type { TPostMedia } from '../types';

export const generatePostMedia = (): TPostMedia => ({
  id: `${Math.random()}`,
  url: 'http://placeimg.com/640/480/cats',
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  tappableObjects: [],
});
