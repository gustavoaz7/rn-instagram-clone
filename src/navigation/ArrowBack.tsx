import React from 'react';
import { useNavigation } from '@react-navigation/native';
import styled, { useTheme } from 'styled-components/native';
import ArrowLeft from '../../assets/svg/arrow-left.svg';

export function ArrowBack(): JSX.Element {
  const navigation = useNavigation();
  const theme = useTheme();

  return (
    <ArrowContainer onPress={navigation.goBack}>
      <ArrowLeft color={theme.color.foreground} />
    </ArrowContainer>
  );
}

const ArrowContainer = styled.Pressable`
  margin-left: ${({ theme }) => theme.spacing.l};
`;
