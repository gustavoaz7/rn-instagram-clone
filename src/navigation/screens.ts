export enum BOTTOM_TAB_SCREENS {
  HOME = '@Bottom-Tab/Home',
  SEARCH = '@Bottom-Tab/Search',
  REELS = '@Bottom-Tab/Reels',
  SHOP = '@Bottom-Tab/Shop',
  PROFILE = '@Bottom-Tab/Profile',
}

export enum DIRECT_STACK_SCREENS {
  DIRECT = '@Direct-Stack/Direct',
  VIDEO = '@Direct-Stack/Video',
  MESSAGE = '@Direct-Stack/Message',
  CHAT = '@Direct-Stack/Chat',
}

export enum HOME_STACK_SCREENS {
  FEED = '@Home-Stack/Feed',
  ACTIVITY = '@Home-Stack/Activity',
  PROFILE = '@Home-Stack/Profile',
}

export enum HOME_SWIPE_SCREENS {
  CAMERA = '@Home-Swipe/Camera',
  BOTTOM_TAB = '@Home-Swipe/Bottom-Tab',
  DIRECT = '@Home-Swipe/Direct',
}

export enum ROOT_STACK_SCREENS {
  HOME_SWIPE = '@Root-Stack/Home-Swipe',
  COMMENTS = '@Root-Stack/Comments',
  STORY = '@Root-Stack/Story',
}
