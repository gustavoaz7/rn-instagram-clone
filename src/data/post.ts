import { TPost } from '../types';
import { generatePostMedia } from './media';

export const createMockPost = (): TPost => ({
  id: `${Math.random()}`,
  owner: {
    id: `${Math.random()}`,
    profilePicUrl: 'https://avatars.githubusercontent.com/u/28453820?v=4',
    username: 'gustavoaz7',
  },
  createdAt: 1620488304326,
  medias: [generatePostMedia()],
  caption:
    'Tempora rerum minima dolor quibusdam blanditiis ea et. Earum iste iure expedita velit laudantium inventore quia qui. Distinctio consectetur accusamus voluptatum hic rem at quas. Incidunt ipsum sed illo molestiae vitae ullam sit consequatur autem. Ut voluptatem nisi. Velit ex dignissimos sed aut.',
  likedBy: ['Tito.Zieme', 'Cleo.Runte69', 'Laverna.Emmerich24'],
  comments: [
    {
      id: `${Math.random()}`,
      owner: {
        id: `${Math.random()}`,
        profilePicUrl: 'https://cdn.fakercloud.com/avatars/antonyzotov_128.jpg',
        username: 'Vivienne_Doyle',
      },
      createdAt: 1620488304326,
      text: 'Aspernatur quod ea unde repellendus magni nihil sunt.',
      likedBy: ['Frankie30', 'Patricia24', 'Westley_Ebert49'],
    },
  ],
  location: 'Daronville',
});
