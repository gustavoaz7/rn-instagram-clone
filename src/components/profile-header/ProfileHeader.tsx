import React, { useCallback, useState } from 'react';
import styled from 'styled-components/native';
import { useNavigation } from '@react-navigation/native';
import { Image, Pressable, ViewProps } from 'react-native';
import { AvatarWithRing } from '../avatar-with-ring';
import { Text } from '../text';
import type { TProfile } from '../../types';
import type { THomeStackNavigationProps } from '../../navigation/HomeStackNavigator';
import { ROOT_STACK_SCREENS } from '../../navigation/screens';

export type TProfileHeaderProps = Pick<
  TProfile,
  'profilePicUrl' | 'postsCount' | 'followCount' | 'followedByCount' | 'story'
> & {
  style?: ViewProps['style'];
};

export const ProfileHeader = ({
  profilePicUrl,
  postsCount,
  followCount,
  followedByCount,
  story,
  style,
}: TProfileHeaderProps) => {
  const [isLoading, setIsLoading] = useState(false);
  const navigation = useNavigation<THomeStackNavigationProps>();

  const handlePress = useCallback(async () => {
    if (!story) return;
    setIsLoading(true);
    await Promise.all(story.medias.map(media => Image.prefetch(media.url)));
    setIsLoading(false);
    navigation.navigate(ROOT_STACK_SCREENS.STORY, { id: story.id, story });
  }, [navigation, story]);

  return (
    <Row style={style} testID="ProfileHeader">
      <Pressable onPress={handlePress}>
        <AvatarWithRing
          size={90}
          imageUrl={profilePicUrl}
          color={story ? 'gradient' : 'transparent'}
          loading={isLoading}
        />
      </Pressable>
      <StatsBar>
        <StatsContainer>
          <Bold>{postsCount}</Bold>
          <Text>Posts</Text>
        </StatsContainer>
        <StatsContainer>
          <Bold>{followedByCount}</Bold>
          <Text>Followers</Text>
        </StatsContainer>
        <StatsContainer>
          <Bold>{followCount}</Bold>
          <Text>Following</Text>
        </StatsContainer>
      </StatsBar>
    </Row>
  );
};

const Row = styled.View`
  flex-direction: row;
`;

const StatsBar = styled(Row)`
  flex: 1;
  justify-content: space-between;
  margin-left: ${({ theme }) => theme.spacing.l};
`;

const StatsContainer = styled.View`
  justify-content: center;
  align-items: center;
`;

const Bold = styled(Text)`
  font-weight: bold;
  font-size: ${({ theme }) => theme.font.size.l};
`;
