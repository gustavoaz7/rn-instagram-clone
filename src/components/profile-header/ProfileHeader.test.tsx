import React from 'react';
import { render, fireEvent, act } from '@testing-library/react-native';
import { Image, Pressable } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { ProfileHeader, TProfileHeaderProps } from './ProfileHeader';
import { Providers } from '../../Providers';
import { generateMockProfile } from '../../data';
import { ROOT_STACK_SCREENS } from '../../navigation/screens';

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));
const useNavigationMock = useNavigation as jest.Mock;

describe('components - ProfileHeader', () => {
  const navigateSpy = jest.fn();
  useNavigationMock.mockReturnValue({ navigate: navigateSpy });
  const props: TProfileHeaderProps = generateMockProfile();
  const options = { wrapper: Providers };

  beforeEach(() => {
    navigateSpy.mockReset();
  });

  it('renders', () => {
    render(<ProfileHeader {...props} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<ProfileHeader {...props} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('inherits style', () => {
    const style = { backgroundColor: 'orange' };
    const { getByTestId } = render(
      <ProfileHeader {...props} style={style} />,
      options,
    );

    expect(getByTestId('ProfileHeader')).toHaveStyle(style);
  });

  it('renders correct profile stats', () => {
    const { queryByText } = render(<ProfileHeader {...props} />, options);

    expect(queryByText(`${props.postsCount}`)).toBeTruthy();
    expect(queryByText(`${props.followedByCount}`)).toBeTruthy();
    expect(queryByText(`${props.followedByCount}`)).toBeTruthy();
  });

  describe('when profile has story', () => {
    it('renders gradient avatar ring', () => {
      const { UNSAFE_getByType } = render(
        <ProfileHeader {...props} />,
        options,
      );

      expect(
        UNSAFE_getByType(Pressable).findByProps({ color: 'gradient' }),
      ).toBeTruthy();
    });

    describe('when avatar is pressed', () => {
      it('prefetches story medias ', async () => {
        const spy = jest.spyOn(Image, 'prefetch');
        const { UNSAFE_getByType } = render(
          <ProfileHeader {...props} />,
          options,
        );

        await act(async () => {
          fireEvent.press(UNSAFE_getByType(Pressable));
        });

        if (!props.story) fail();
        expect(spy).toHaveBeenCalledTimes(props.story.medias.length);
        props.story.medias.forEach((media, i) => {
          expect(spy).toHaveBeenNthCalledWith(i + 1, media.url);
        });
        spy.mockRestore();
      });

      it('navigates to story screen on press', async () => {
        const { UNSAFE_getByType } = render(
          <ProfileHeader {...props} />,
          options,
        );

        await act(async () => {
          fireEvent.press(UNSAFE_getByType(Pressable));
        });

        expect(navigateSpy).toHaveBeenCalledTimes(1);
        expect(navigateSpy).toHaveBeenCalledWith(ROOT_STACK_SCREENS.STORY, {
          id: props.story?.id,
          story: props.story,
        });
      });
    });
  });

  describe('when profile has no story', () => {
    const noStoryProps: TProfileHeaderProps = { ...props, story: undefined };

    it('renders transparent avatar ring', () => {
      const { UNSAFE_getByType } = render(
        <ProfileHeader {...noStoryProps} />,
        options,
      );

      expect(
        UNSAFE_getByType(Pressable).findByProps({ color: 'transparent' }),
      ).toBeTruthy();
    });

    describe('when avatar is pressed', () => {
      it('does not prefetches any image ', async () => {
        const spy = jest.spyOn(Image, 'prefetch');
        const { UNSAFE_getByType } = render(
          <ProfileHeader {...noStoryProps} />,
          options,
        );

        await act(async () => {
          fireEvent.press(UNSAFE_getByType(Pressable));
        });

        expect(spy).toHaveBeenCalledTimes(0);
        spy.mockRestore();
      });

      it('does not navigates to story screen on press', async () => {
        const { UNSAFE_getByType } = render(
          <ProfileHeader {...noStoryProps} />,
          options,
        );

        await act(async () => {
          fireEvent.press(UNSAFE_getByType(Pressable));
        });

        expect(navigateSpy).toHaveBeenCalledTimes(0);
      });
    });
  });
});
