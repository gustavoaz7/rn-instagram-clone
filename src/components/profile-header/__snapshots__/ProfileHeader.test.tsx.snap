// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - ProfileHeader matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
    ]
  }
  testID="ProfileHeader"
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View
      size={90}
      style={
        Array [
          Object {
            "borderRadius": 90,
            "height": 90,
            "overflow": "hidden",
            "position": "relative",
            "width": 90,
          },
        ]
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              4292948076,
              4294750021,
            ],
            "endPoint": Array [
              0.3,
              0.7,
            ],
            "locations": undefined,
            "startPoint": Array [
              0.7,
              0.3,
            ],
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <View
          ringWidth={2}
          size={90}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 86,
                "height": 86,
                "justifyContent": "center",
                "width": 86,
              },
            ]
          }
        >
          <Image
            offset={2}
            ringWidth={2}
            size={90}
            source={
              Object {
                "uri": "https://avatars.githubusercontent.com/u/28453820?v=4",
              }
            }
            style={
              Array [
                Object {
                  "borderColor": "#26262633",
                  "borderRadius": 82,
                  "borderWidth": 2,
                  "height": 82,
                  "width": 82,
                },
              ]
            }
          />
        </View>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexDirection": "row",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
          "marginLeft": 20,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
            },
          ]
        }
      >
        20
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
          ]
        }
      >
        Posts
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
            },
          ]
        }
      >
        50
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
          ]
        }
      >
        Followers
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "bold",
            },
          ]
        }
      >
        47
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#262626",
            },
          ]
        }
      >
        Following
      </Text>
    </View>
  </View>
</View>
`;
